### PR TITLE
JDK-8207369 Add more robot tests.

### PR DESCRIPTION
**Pull request by brcolow**
(Partially? at least a step in the right direction...) resolves robot API follow-up issue [JDK-8207369](https://bugs.openjdk.java.net/browse/JDK-8207369).

### CODE REVIEW

File: tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java [hunk 1]
```diff
@@ -133,26 +258,135 @@ public void testMouseMove() {
...
+
+        @Override
+        public void handle(MouseEvent event) {
+            String expectedText = mouseAction + " " + mouseButton;
+            if (event.getButton() == mouseButton) {
+                button.setText(expectedText);
+                onClickLatch.countDown();
+            } else {
+                button.setText(String.format(mouseAction + " wrong button (expected \"%s\" but got \"%s\")",
+                        mouseButton, event.getButton()));
```
[the remark this answers is not in the document]
Comment (brcolow) [reply tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java]: Good catch. I will investigate the many failures.

**Comment by brcolow** [comment]
I think the failures (besides the mouse scroll amount which seems to be hardcoded to a value of (-)40 on Linux) are happening because you can't release "in a vacuum" without a preceding press. I will revert those to testing mouse clicks (with press and release).

**Comment by brcolow** [comment]
The mouse scroll wheel failure is kind of interesting. Basically the GtkRobot sends multiple fake mouse buttons presses (4 for scroll up, 5 for scroll down), one for each int of the `wheelAmt`. Therefore I think what I can do is add a sort of timed collection to accumulate all scroll events sent to the `Button` and then add them up. I may still need to mess with the multiplier though on a system-dependent basis which is going to be ugly. The reason for this is that the JavaFX event handler will get 5 scroll events each with a scroll amount of (+/-) 40 for a total of 200. Then one would need to divide by the multiplier to match the passed-in `wheelAmt`.

**Comment by brcolow** [comment]
For some reason `event.getTotalDeltaY()` reports 0. But I was able to hack something together that passes on Windows and Linux. Still need to test macOS.

**Comment by brcolow** [comment]
I finally was able to test and fix the new tests on macOS. I have now confirmed that the new tests pass on all 3 operating systems. I am not really happy with the approach I took in order to get a unified scroll amount between platforms but wasn't sure of an alternative approach.

**Comment by brcolow** [comment]
The new drag tests pass locally for me on Windows but seem to fail in a Linux virtual machine but I can't tell if that's because it's really slow or because there is an actual problem. Do you have a local Linux install you could test them on?

**Comment by kevinrushforth** [comment]
I can test them later this morning when I get into the office.

**Comment by kevinrushforth** [comment]
The tests all PASS on my Linux system (a physical machine running Ubuntu 18.04). I'll run on Windows and Mac and then finish my review.

I don't have a good answer to your question about the platform differences regarding the scrolling multiplier (-1 on Mac versus 40 on Windows and Linux). The negative versus positive might be related to the "natural scrolling order" on Mac, which is intentionally backwards by default from the other platforms. However, a difference of 40x in magnitude points to something wrong in the implementation.

File: tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java [hunk 2]
```diff
@@ -167,23 +372,124 @@ public void testMouseClick() {
...
+    }
+
+    private static void testMouseWheel(int amount) {
         CountDownLatch onScrollLatch = new CountDownLatch(1);
         CountDownLatch setSceneLatch = new CountDownLatch(1);
         Button button = new Button("Scroll me");
         InvalidationListener invalidationListener = observable -> setSceneLatch.countDown();
+        int[] totalScroll = new int[]{0};
+        long[] firstScrollMillis = new long[]{0};
+        int scrollMultiplier = System.getProperty("os.name").toLowerCase().startsWith("mac") ? -1 : 40;
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java]: Can you add a comment here? This suggests that there might be an issue with scrolling in the Mac robot implementation. We might want to file a follow-up issue to investigate.

**Comment by brcolow** [comment]
Added a commit which addresses both issues.

File: tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java [hunk 3]
```diff
@@ -474,10 +474,12 @@ private static void testMouseWheel(int amount) {
         InvalidationListener invalidationListener = observable -> setSceneLatch.countDown();
         int[] totalScroll = new int[]{0};
         long[] firstScrollMillis = new long[]{0};
+        // The scroll wheel amount is multiplied by 40 on Linux and Windows, but not on macOS. The
+        // directions are also reversed.
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java]: I filed JDK-8214580 to investigate this. Can you add to this comment that the difference in behavior is unexpected and might be a bug? Something like:

```
// This difference in behavior is unexplained and might be a bug. See JDK-8214580.
```

File: tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java [hunk 4]
```diff
@@ -474,10 +474,12 @@ private static void testMouseWheel(int amount) {
...
         int scrollMultiplier = System.getProperty("os.name").toLowerCase().startsWith("mac") ? -1 : 40;
         Util.runAndWait(() -> {
             button.setOnScroll(event -> {
-                totalScroll[0] += event.getDeltaY();
+                totalScroll[0] += event.getDeltaY() * Screen.getPrimary().getOutputScaleY();
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java]: Looks good. I'll test it on Monday.

File: tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java [hunk 5]
```diff
@@ -474,10 +474,13 @@ private static void testMouseWheel(int amount) {
...
         int[] totalScroll = new int[]{0};
         long[] firstScrollMillis = new long[]{0};
+        // The scroll wheel amount is multiplied by 40 on Linux and Windows, but not on macOS. The
+        // directions are also reversed. This difference in behavior is unexplained and might be a bug.
+        // See JDK-8214580.
         int scrollMultiplier = System.getProperty("os.name").toLowerCase().startsWith("mac") ? -1 : 40;
         Util.runAndWait(() -> {
             button.setOnScroll(event -> {
-                totalScroll[0] += event.getDeltaY();
+                totalScroll[0] += event.getDeltaY() * Screen.getPrimary().getOutputScaleY();
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java]: Multiplying by the output scale works correctly on Windows. It doesn't work on a Mac with a retina display, though. The test now fails for me. This suggests another platform difference between how Robot::mouseWheel works on Mac versus the other platforms.

Comment (brcolow) [reply tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java]: Do you think we should skip the test on Mac for now and un-skip it as part of JDK-8214580?

Comment (brcolow) [reply tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java]: Okay, I added a new commit and updated the bug.

**Comment by kevinrushforth** [comment]
That looks good. One quick thought: for consistency, the following could also use `PlatformUtil.isMac()` now that the test imports it:

```
    int scrollMultiplier = System.getProperty("os.name").toLowerCase().startsWith("mac") ? -1 : 40;
```

I'll do a quick final test on Mac and Windows and then I think we can get this in.

**Comment by kevinrushforth** [comment]
Btw, this would have been easier for me to review without the squash / force-push

**Comment by brcolow** [comment]
Ah, sorry about that. Will keep that in mind going forward.

**Pull request merged**